Documentation on using OCI images for model storage (modelcars) (#415)
opendatahub-io/kserve: docs/odh/oci-model-storage.md

# Using OCI containers for model storage

> Starting ODH 2.18, the ability to use OCI containers as storage for models

Reviewer: ```suggestion
Starting from ODH 2.18, the ability to use OCI containers as storage for models
```

is enabled in KServe by default. The benefits of using OCI containers for
model storage are described in [the upstream KServe project

> documentation](https://kserve.github.io/website/latest/modelserving/storage/oci/),

Reviewer 2: maybe better to link only in the _documentation_ word.

> which also describes how to deploy models from OCI images. 

Reviewer: ```suggestion
which also explains how to deploy models from OCI images. 
```

This page offers a guide similar to the upstream project documentation, but
focusing on the OpenDataHub and OpenShift characteristics. To demonstrate

> how to create an OCI image, the publicly available [MobileNet v2-7

Reviewer 2: ```suggestion
how to create an OCI container image, the publicly available [MobileNet v2-7
```

model](https://github.com/onnx/models/tree/main/validated/vision/classification/mobilenet)
is used. This model is in ONNX format.

The ODH projects provides configurations for the OpenVINO model server, which

> supports models in ONNX format. Thus, this guide will use this model server

Reviewer 2: Maybe it is worth mentioning that OpenVINO does not support all ONNX formats, if I am not mistaken, only DL.

Reviewer: OpenVino supports ONNX. @Xaenalt please confirm that.

Author: > Maybe it is worth mentioning that OpenVINO does not support all ONNX formats

I don't know. This doc is not meant to be detailed about OpenVINO.

to demonstrate how deploy the MobileNet v2-7 model stored in an OCI image.

## Storing a model in an OCI image

Start by creating an empty directory for downloading the model and creating
the necessary support files to create the OCI image. You may use a temporary
directory by running the following command:
```shell
cd $(mktemp -d)
```

OpenVINO expects a specific directory tree for model versioning.
Starting from some base directory, its contents should be a collection of
numbered subdirectories using positive integer values. The numbers would

> represent the versions of the model. When using OCI images, this

Reviewer 2: ```suggestion
represent the versions of the model. When using OCI container images, this
```

structure may be irrelevant, as you can use the OCI container registry

> features. However, since OpenVINO expects the versioned directory structure, a

Reviewer 2: what features are you referring to?

Author: For example, typically image tags are used for versions.
But, ultimately, it depends how the user is using image repositories and tags. So, I didn't want to mention anything.

single subdirectory with an artificial version `1` can be used. Using `models/` as the
base path, create the expected directory structure and download the sample

> model to it:

Reviewer: ```suggestion
model into it:
```

```shell

> # 

Reviewer 2: ```suggestion

```

mkdir -p models/1

DOWNLOAD_URL=https://github.com/onnx/models/raw/main/validated/vision/classification/mobilenet/model/mobilenetv2-7.onnx
curl -L $DOWNLOAD_URL -O --output-dir models/1/
```

> [!TIP]
> If you are planning to use a different model server, you should adapt this

> > guide accordingly to your model server requirements. Typically, you would need

Reviewer 2: ```suggestion
> guide accordingly to your requirements. Typically, you would need
```

Reviewer: `model server requirements` make sense to me

Reviewer 2: just to avoid redundancy, this sentence is about `model server`.

Author: This time, I prefer the redundancy.

> to place your model files directly under `models/`.

Create a file named `Containerfile` with the following contents:
```Dockerfile

> FROM registry.access.redhat.com/ubi8/ubi-micro:latest

Reviewer 3: just curious do we use
registry.access.redhat.com  images for upstream ?

Reviewer: It is a UBI image, so the image should be correct, as it does not require credentials for use.

Author: > registry.access.redhat.com images for upstream ?

Not in upstream KServe project. There, `busybox` is being used: https://hub.docker.com/_/busybox
In OpenShift ecosystem, I think UBI is preferred.

> FROM registry.access.redhat.com/ubi8/ubi-micro:latest

Reviewer 2: better to use registry.redhat.io

Reviewer 2: oh, yeah, correct :)

> COPY --chown=0:0 models /models

Reviewer: as @spolti mentioned, I don't understand why you want to use root for the directory?

Author: Read the reply: https://github.com/opendatahub-io/kserve/pull/415/files#r1788160657

Reviewer: I disagree with your point. It is not just about not using root for executing processes; any file or folder included in the pod should have permissions that only allow a user or group to access it. Granting permissions that allow anyone to use those files or folders poses significant security risks. Therefore, I believe we need to change the owner/group of all files/folders we need to use to a user other than root

Reviewer: It gives only read permission for group of the specific models and added USER so I think it should be ok for now.

RUN chmod -R a=rX /models
```

Notice that model files are copied into `/models` inside the container. KServe
expects this path to exist in the OCI image and also expects the model files to
be inside it.

> Also, notice that `ubi8-micro` is used as a base image. Empty images, like

Reviewer 2: ```suggestion
Also, notice that `ubi8-micro` is used as a base container image. Empty images, like
```

`scratch` cannot be used, because KServe needs to configure the model image
with a command to keep it alive and ensure the model files remain available in
the pod. Thus, it is required to use a base image that provides a shell.

Finally, notice that ownership of the copied model files is changed to the `root`
user and group, and also read permissions are granted to all users. This is
important, because OpenShift runs containers with a random user ID and with the

> `root` group ID. The adjustment of the group and the privileges on the model files

Reviewer 2: I am, not sure about it, avoiding user root is a good practice.
Did you try to add a dummy user and use this uid instead 0?

e.g.: `COPY --chown=1001:0 models /models`

Author: My understanding is that the advice against using `root` is applicable for running processes. In that case, you definitely want to avoid running as `root` because it has security implications.

In contrast, here the models are simple data files and cannot be run. The `chmod` command in the Containerfile only gives read permissions and won't give write nor execute permissions. The model server (which is another container) would load the model and run it, and would be running as non-root user (provided the user configured the `ServingRuntime` correctly).

So, given the models are simple data files, I'm not seeing security implications with giving root ownership. It  has the added benefit that the model files cannot be modified once mounted in the pod. So, I preferred to keep the Containerfile simple and not creating users, nor using dummy UIDs.

Author: I think I'm not understanding the relationship between bash and the data files... And the base image has a lot of files owned by root. Probably, I'm not seeing something as I don't understand the problem.

Reviewer 2: It is more about good practices than a real problem.

Reviewer: I agreed with @spolti

ensures that the model server can access them.

Verify that the directory structure is good using the `tree` command:
```shell
tree

.
├── Containerfile
└── models
    └── 1
        └── mobilenetv2-7.onnx
```

> [!NOTE]
> Remember that the shown directory structure under `models/` is specific to OpenVINO.

> Create the OCI container image with podman, and upload it to a registry. For

Reviewer 2: ```suggestion
Create the OCI container image with Podman, and upload it to a registry. For
```

example, using Quay as the registry:
```shell
podman build --format=oci -t quay.io/<user_name>/<repository_name>:<tag_name> .
podman push quay.io/<user_name>/<repository_name>:<tag_name>
```

> [!TIP]

> > When uploading your image, if your image repository is private, ensure you

Reviewer 2: ```suggestion
> When uploading your container image, if your repository is private, ensure you
```

caikit-tgis-serving-template         Caikit is an AI toolkit that enables users to manage models through a set of...    0 (all set)   1
kserve-ovms                          OpenVino Model Serving Definition                                                  0 (all set)   1
ovms                                 OpenVino Model Serving Definition                                                  0 (all set)   1
tgis-grpc-serving-template           Text Generation Inference Server (TGIS) is a high performance inference engin...   0 (all set)   1
vllm-runtime-template                vLLM is a high-throughput and memory-efficient inference and serving engine f...   0 (all set)   1
```

The template that is applicable for KServe and holds the OpenVINO configuration
is the one named as `kserve-ovms`. To create an instance of it, run the
following command:
```shell
oc process -n opendatahub -o yaml kserve-ovms | oc apply -f -
```

You can verify that the `ServingRuntime` has been created successfully with the
following command:
```shell
oc get servingruntimes

NAME          DISABLED   MODELTYPE     CONTAINERS         AGE
kserve-ovms              openvino_ir   kserve-container   1m
```

> Notice that the ServingRuntime has been created with `kserve-ovms` name.

Reviewer: ```suggestion
> [!NOTE]
> The ServingRuntime has been created with `kserve-ovms` name.
```

Author: Rather than being a side note, I wanted to imply that the user should "take note" of the name, as it is used on the `InferenceService`... So, not sure if it should be quoted as a `[!NOTE]`.

Reviewer: my bad, I thought that it was `NOTE` not `Notice`

metadata:
  name: sample-isvc-using-oci
spec:
  predictor:
    model:
      runtime: kserve-ovms # This is the name of the ServingRuntime resource
      modelFormat:
        name: onnx
      storageUri: oci://quay.io/<user_name>/<repository_name>:<tag_name>
```

> [!IMPORTANT]
> The resulting `ServingRuntime` and `InferenceService` configurations won't set
> any resource limits.

Once the `InferenceService` resource is created, KServe will deploy the model
stored in the OCI image referred by the `storageUri` field. Check the status
of the deployment with the following command:
```shell
oc get inferenceservice

NAME                    URL                                                       READY   PREV   LATEST   PREVROLLEDOUTREVISION   LATESTREADYREVISION                     AGE
sample-isvc-using-oci   https://sample-isvc-using-oci-oci-model-example.example   True           100                              sample-isvc-using-oci-predictor-00001   1m
```

> [!IMPORTANT]
> Remember that, by default, models are exposed outside the cluster and not
> protected with authorization. Read the [authorization guide](authorization.md#deploying-a-protected-inferenceservice)

> > and the [private services guide (TODO)](#TODO) to learn how to privately deploy

Reviewer 2: what is missing in this TODO?

Reviewer 2: Maybe is good to also create the jiras so we don't forgot about it.

> models and how to protect them with authorization.

## Deploying a model stored in an OCI image from a private repository

To deploy a model stored in a private OCI repository you need to configure an
image pull secret. For detailed documentation, please [consult the OpenShift

> documentation for image pull secrets](https://docs.openshift.com/container-platform/latest/openshift_images/managing_images/using-image-pull-secrets.html).

Reviewer 2: maybe link just the `documentation` word

When using namespaced pull secrets you can create a pull secret using the following
command template:

```shell
oc create secret docker-registry <pull-secret-name> \
  --docker-server=<registry-server> \
  --docker-username=<username> \
  --docker-password=<password>
```

> Once the pull secret is created, you can follow the steps from the previous

Reviewer 2: what is the previous section?

Author: i.e. the previous heading _Deploying a model stored in an OCI image in a public repository_

Reviewer 2: I guess I didn't find it

Reviewer 2: oops.
